killswitch.py

```diff
@@ -1,54 +1,231 @@
+
+        if isinstance(current, Mapping):
+            current = current[key]  # type: ignore # I really dont know at this point what you want from me mypy.
+
+        elif isinstance(current, Sequence):
+            target_idx = _get_int(key)  # mypy is broken. doesn't like := here.
+            if target_idx is not None:
+                current = current[target_idx]
+            else:
+                raise ValueError(f'Cannot index sequence with non-int key {key!r}')
+
+        else:
+            raise ValueError(f'Dont know how to index a {type(current)} ({current!r})')
+
+    _apply(current, path, to_set, delete)
+
+
+def _get_int(s: str) -> Optional[int]:
+    try:
+        return int(s)
+    except ValueError:
+        return None
+
+
+class KillSwitches(NamedTuple):
     """One version's set of kill switches."""
 
     version: semantic_version.SimpleSpec
-    kills: Dict[str, str]
+    kills: Dict[str, SingleKill]
+
+    @staticmethod
+    def from_dict(data: KillSwitchSetJSON) -> KillSwitches:
+        """Create a KillSwitches instance from a dictionary."""
+        ks = {}
+
+        for match, ks_data in data['kills'].items():
+            ks[match] = SingleKill(
+                match=match,
+                reason=ks_data['reason'],
+                redact_fields=ks_data.get('redact_fields'),
+                set_fields=ks_data.get('set_fields'),
+                delete_fields=ks_data.get('delete_fields')
+            )
+
+        return KillSwitches(version=semantic_version.SimpleSpec(data['version']), kills=ks)
 
 
 class DisabledResult(NamedTuple):
     """DisabledResult is the result returned from various is_disabled calls."""
 
     disabled: bool
-    reason: str
+    kill: Optional[SingleKill]
+
+    @property
+    def reason(self) -> str:
+        """Reason provided for why this killswitch exists."""
+        return self.kill.reason if self.kill is not None else ""
+
+    def has_kill(self) -> bool:
+        """Return whether or not this DisabledResult has a Kill associated with it."""
+        return self.kill is not None
+
+    def has_rules(self) -> bool:
+        """Return whether or not the kill on this Result contains rules."""
+        # HACK: 2021-07-09 # Python/mypy/pyright does not support type guards like this yet. self.kill will always
+        # be non-None at the point it is evaluated
+        return self.has_kill() and self.kill.has_rules  # type: ignore
 
 
 class KillSwitchSet:
     """Queryable set of kill switches."""
 
-    def __init__(self, kill_switches: List[KillSwitch]) -> None:
+    def __init__(self, kill_switches: List[KillSwitches]) -> None:
         self.kill_switches = kill_switches
 
-    def get_disabled(self, id: str, *, version=_current_version) -> DisabledResult:
+    def get_disabled(self, id: str, *, version: Union[Version, str] = _current_version) -> DisabledResult:
         """
         Return whether or not the given feature ID is disabled by a killswitch for the given version.
 
```

Comment:
flake8: no blank line after function docstring (yes, I hate the inconsistency with *requiring* such after class docstrings).